packages/middleware/src/loggerManager.ts

```diff
@@ -0,0 +1,108 @@
+import type { Response } from "express";
+import { LoggerInterface } from "@vue-storefront/logger";
+import type {
+  AlokaiContainer,
+  IntegrationContext,
+  MiddlewareConfig,
+} from "./types";
+
+type LoggerInstance = LoggerInterface;
+const GLOBAL_MIDDLEWARE_CFG_KEY = Symbol("GLOBAL_MIDDLEWARE_CFG_KEY");
+export class LoggerManager<
+  TLoggerConfig = any,
+  TIntegrationContext extends Record<string, IntegrationContext> = any
+> {
+  private instances: Record<string | symbol, LoggerInstance> = {};
+
+  /** Creating minimal amount of needed Logger instances * */
+  constructor(
+    config: MiddlewareConfig<TIntegrationContext>, // Whole middleware.config.js object
+    buildLogger: (logger: TLoggerConfig) => LoggerInstance
+  ) {
+    // Global Logger
+    this.instances[GLOBAL_MIDDLEWARE_CFG_KEY] = buildLogger(
+      config.logger as TLoggerConfig // logger is on same level as integrations
+    );
+    // Logger for every integration having custom logger config
+    for (const [integrationName, integrationEntry] of Object.entries(
+      config.integrations
+    )) {
+      if (this.hasCustomLoggerConfig(integrationEntry)) {
+        // every integration inside `configuration` object can have own logger too
+        this.instances[integrationName] = buildLogger(
+          integrationEntry.logger as TLoggerConfig
+        );
+      }
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  private hasCustomLoggerConfig(integrationEntry) {
+    return Boolean(integrationEntry.logger);
+  }
+
+  // getting reference to instance of logger by integration tag name
+  public get(integrationName) {
+    return (
+      this.instances[integrationName] ||
+      this.instances[GLOBAL_MIDDLEWARE_CFG_KEY]
+    );
+  }
+
+  // getting reference to global instance of logger
+  public getGlobal() {
+    return this.instances[GLOBAL_MIDDLEWARE_CFG_KEY];
+  }
+}
+
+export function getLogger(
```

Comment:
```
export function getLogger(
  source: AlokaiContainer | Response | { res: Response }
): LoggerInstance {
  if (source?.logger) {
    return source.logger;
  }

  const base = source.res || source;

  if (base?.locals?.alokai?.logger) {
    return base.locals.alokai.logger;
  }

  throw new Error("Logger instance could not be determined");
}
```

Reply:
Or instead of throw new Error use console.log as a fallback but add log info about the problem

Reply:
I prefer an approach with throwing. The idea with fallback requires mapping different severity of our logger's methods to console.log/warn/error etc. The probability of that scenario is near 0% and even if it happens, app won't start, so it won't be deployed and information from error will be visible.

Reply:
As a next step of [this approach](https://github.com/vuestorefront/vue-storefront/pull/7269#discussion_r1791770263), we could do smth like:
```ts
type ContextSubset = { res: Response };
type LoggerSource = AlokaiContainer | Response | ContextSubset;
function isAlokaiContainer(source: LoggerSource): source is AlokaiContainer {
  return "logger" in source;
}
function isContext(source: LoggerSource): source is ContextSubset {
  return "res" in source;
}
function findLogger(source: LoggerSource): LoggerInterface {
  if (isAlokaiContainer(source)) {
    return source?.logger;
  }
  const base = isContext(source) ? source?.res : source;
  return base?.locals?.alokai?.logger;
}
export function getLogger(source: LoggerSource): LoggerInterface {
  const logger = findLogger(source);
  if (!logger) {
    throw new Error("Logger instance could not be determined");
  }
  return logger;
}
```